api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,155 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// RateLimitFilter allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type RateLimitFilter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of RateLimitFilter.
+	Spec RateLimitFilterSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitFilterList contains a list of RateLimitFilter resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimitFilter `json:"items"`
+}
+
+// RateLimitFilterSpec defines the desired state of RateLimitFilter.
+// +union
+type RateLimitFilterSpec struct {
+	// Type decides the scope for the RateLimits.
+	// Valid RateLimitType values are:
+	//
+	// * "Global" - In this mode, the rate limits are applied across all Envoy proxy instances.
+	//
+	// +unionDiscriminator
+	Type RateLimitType `json:"type"`
+	// Global rate limit configuration.
+	//
+	// +optional
+	Global *GlobalRateLimit `json:"global"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	GlobalRateLimitType RateLimitType = "Global"
+)
+
+// GlobalRateLimit defines the global rate limit configuration.
+type GlobalRateLimit struct {
+	// Rules are a list of RateLimit matchers and limits.
+	// Each rule and its associated limit is applied
```

Comment:
wording suggestion/assumption on what youre trying to convey here:

```
Multiple rules may match a request. The effective rate limit applied to a request is the computed aggregate of the rate limits configured in these matching rules.
```

Once there are docs etc. might be good to point to something giving an example, e.g. if you have two rules with header value matches that overlap, the stricter rate limit will win

Reply:
@sunjayBhatia this understanding is incorrect, if there are multiple rules e.g.
```
rules:
- matches:
  - headers
       name: :method
       value: GET
  limit:
    requests: 2
    unit: SECOND
- matches:
  - headers
       name: :authority
       value: example.com
  limit:
    requests: 2
    unit: HOUR          
```

lets assume match holds true, both will get applied simultaneously in a mutually exclusive way

Reply:
the difference between a HTTPRoute rule is, the request must go to some/one backend, so the top level is ORed`, here the RateLimit can apply to any/all

Reply:
if one of those was 2/hour and the other 4/hour, which one would be applied to the 3rd GET to example.com?

Reply:
both would i.e. the 4 requests would get applied to each of the `rule`s, outcome is same, 2nd and 4th request gets rate limited, but the count is incremented for each rule

Reply:
the point i was making above is for example with `header value matches that overlap`, if you have match on `:method` AND `:authority` and a separate rule to match on `:authority`, if you somehow configured a stricter limit on the only `:authority` rule, that will have precedence when the limits are evaluated even if both buckets are counted separately

Reply:
precedence is used to break ties, there is no notion of ties here, all buckets are counted and individually evaluated

Reply:
similar to how more specific matches or longer path prefixes etc. are documented to "win" in route matching, some documentation on what happens when you have multiple limits that apply to a request would be good to include

Reply:
this case captures what I'm saying: https://github.com/envoyproxy/gateway/pull/706#discussion_r1053594174

users should be notified that e.g. Alice and Bob will still be rate limited at 10RPS because that limit is more strict than what they think they may have configured for those specific users, the buckets are counted separately yes but when evaluating if a request has gone over any limits they are aggregated

Reply:
Even before there are end-user docs, if the design goal is that the strictest match wins, the design doc should say that.

Reply:
but the goal is not strictest match wins, its any & all match win

Reply:
yes, but effectively that means that if requests match multiple limiting rules, the buckets will fill up for the stricter limit first so users will see their requests rate limited according to that limit first

in practice I have seen users get very confused as to how things work when such overlapping rate limits are in place, thinking they shouldnt get rate limited on certain requests when their config says they should